providers/ldap/ldap_client.go

```diff
@@ -0,0 +1,655 @@
+package ldap
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/rancher/go-rancher/v2"
+	"github.com/rancher/rancher-auth-service/model"
+	"gopkg.in/ldap.v2"
+)
+
+// LClient is the ldap client
+type LClient struct {
+	Config            *model.LdapConfig
+	ConstantsConfig   *ConstantsConfig
+	SearchConfig      *SearchConfig
+	AccessMode        string
+	AllowedIdentities string
+}
+
+type SearchConfig struct {
+	Server               string
+	Port                 int64
+	BindDN               string
+	BindPassword         string
+	UserSearchAttributes []string
+	GroupSeachAttributes []string
+}
+
+type ConstantsConfig struct {
+	UserScope            string
+	GroupScope           string
+	Scopes               []string
+	MemberOfAttribute    string
+	ObjectClassAttribute string
+	LdapJwt              string
+}
+
+var nilIdentity = client.Identity{Resource: client.Resource{
+	Type: "identity",
+}}
+var nilToken = model.Token{Resource: client.Resource{
+	Type: "token",
+}}
+
+func (l *LClient) InitializeSearchConfig() *SearchConfig {
+	c := l.ConstantsConfig
+	return &SearchConfig{
+		Server:       l.Config.Server,
+		Port:         l.Config.Port,
+		BindDN:       l.Config.ServiceAccountUsername,
+		BindPassword: l.Config.ServiceAccountPassword,
+		UserSearchAttributes: []string{c.MemberOfAttribute,
+			c.ObjectClassAttribute,
+			l.Config.UserObjectClass,
+			l.Config.UserLoginField,
+			l.Config.UserNameField,
+			l.Config.UserSearchField,
+			l.Config.UserEnabledAttribute},
+		GroupSeachAttributes: []string{c.MemberOfAttribute,
+			c.ObjectClassAttribute,
+			l.Config.GroupObjectClass,
+			l.Config.UserLoginField,
+			l.Config.GroupNameField,
+			l.Config.GroupSearchField},
+	}
+}
+
+func (l *LClient) newConn() (*ldap.Conn, error) {
+	log.Debug("Now creating Ldap connection")
+	searchConfig := l.SearchConfig
+	lConn, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", searchConfig.Server, searchConfig.Port))
+	if err != nil {
+		return nil, fmt.Errorf("Error %v creating connection", err)
+	}
+	lConn.SetTimeout(time.Duration(l.Config.ConnectionTimeout) * time.Second)
+
+	err = lConn.Bind(l.Config.ServiceAccountUsername, l.Config.ServiceAccountPassword)
+	if err != nil {
+		return nil, fmt.Errorf("Error %v in ldap bind", err)
+	}
+
+	return lConn, nil
+}
+
+// GenerateToken generates token
+func (l *LClient) GenerateToken(jsonInput map[string]string) (model.Token, error) {
+	log.Info("Now generating Ldap token")
+	searchConfig := l.SearchConfig
+
+	//getLdapToken:ADTokenCreator
+	//getIdentities: ADIdentityProvider
+
+	split := strings.Split(jsonInput["code"], ":")
+	username, password := split[0], split[1]
+	externalId := username
+	if l.Config.LoginDomain != "" {
+		externalId = l.Config.LoginDomain + "\\" + username
```

Comment:
What if username.contains("\\") already. Seems like we should add the LoginDomain only when the username does not contain "\\"